### DISCUSSION TITLE
Set script content with curl

### DISCUSSION THREAD

**Discussion by gith4me** [arendst/Tasmota#14302] [Q&A] [answered]
Is it possible to set the script content by using curl ?

**Comment by barbudor** [comment]
It should be possible to POST as a form data as the script editor is nothing more than a form with

```
curl -v -"http://192.168.168.195/ta" -F "c1=" -F "save=" -F "t1=>D hello world"
```
works but is not much usefull

```
curl -v -"http://192.168.168.195/ta" -F "c1=" -F "save=" -F "t1=@script.txt"
```
is rejected with a "script memory error"
although everything looks ok wiht Wireshark

May be the ESP webserver doesn't like `filename="script.txt"` in 
```
--------------------------11c6d42c3ed53f2b
Content-Disposition: form-data; name="t1"; filename="script.txt"
Content-Type: text/plain
```

May be @gemu2015 has the solution ?

**Reply by gemu2015** [reply]
sorry i can not help here.
i am not very familiar with curl but tried without success to transfer a script. i can't figure out where the problem is.
unfortunately Wireshark  does not like my WLAN chip. most of the time it immediately blocks any WLAN traffic and even after termination my WLAN is dead. i have to restart to get WLAN working again.
i tried several Wireshark version without success.

in my external script editor i transfer a script in a simple POST action without any problems. 
i use the following to POST a script.txt to http://192.168.178.197/exs?execute_script

using this objective c call:

 urlRequest = [self postRequestWithURL:serverURL boundry:BOUNDRY data:[filePath dataUsingEncoding:NSUTF8StringEncoding]];

i assumed curl would not be a problem, but without Wireshark i can't find the reason what goes wrong.

**Reply by barbudor** [reply]
Interresting
I was trying to use the `/ta` path like the built-on web editor
I will try again with `/exs?execute_script` later today

**Comment by gith4me** [comment]
Tried the above suggestion:

```
curl -H "Content-Type:text/plain" --data $'>D\r\nx=77\r\n>B\r\n=>power 1\r\n' http://s26-001/exs?execute_script 
```

Trace of curl:

```
== Info:   Trying 192.168.10.50:80...
== Info: Connected to s26-001 (192.168.10.50) port 80 (#0)
=> Send header, 135 bytes (0x87)
0000: 50 4f 53 54 20 2f 65 78 73 3f 65 78 65 63 75 74 POST /exs?execut
0010: 65 5f 73 63 72 69 70 74 20 48 54 54 50 2f 31 2e e_script HTTP/1.
0020: 31 0d 0a 48 6f 73 74 3a 20 73 32 36 2d 30 30 31 1..Host: s26-001
0030: 0d 0a 55 73 65 72 2d 41 67 65 6e 74 3a 20 63 75 ..User-Agent: cu
0040: 72 6c 2f 37 2e 38 31 2e 30 0d 0a 41 63 63 65 70 rl/7.81.0..Accep
0050: 74 3a 20 2a 2f 2a 0d 0a 43 6f 6e 74 65 6e 74 2d t: */*..Content-
0060: 54 79 70 65 3a 74 65 78 74 2f 70 6c 61 69 6e 0d Type:text/plain.
0070: 0a 43 6f 6e 74 65 6e 74 2d 4c 65 6e 67 74 68 3a .Content-Length:
0080: 20 32 35 0d 0a 0d 0a                             25....
=> Send data, 25 bytes (0x19)
0000: 3e 44 0d 0a 78 3d 37 37 0d 0a 3e 42 0d 0a 3d 3e >D..x=77..>B..=>
0010: 70 6f 77 65 72 20 31 0d 0a                      power 1..
== Info: Mark bundle as not supporting multiuse
<= Recv header, 24 bytes (0x18)
0000: 48 54 54 50 2f 31 2e 31 20 33 30 33 20 53 65 65 HTTP/1.1 303 See
0010: 20 4f 74 68 65 72 0d 0a                          Other..
<= Recv header, 25 bytes (0x19)
0000: 43 6f 6e 74 65 6e 74 2d 54 79 70 65 3a 20 74 65 Content-Type: te
0010: 78 74 2f 68 74 6d 6c 0d 0a                      xt/html..
<= Recv header, 16 bytes (0x10)
0000: 4c 6f 63 61 74 69 6f 6e 3a 20 2f 65 78 73 0d 0a Location: /exs..
<= Recv header, 19 bytes (0x13)
0000: 43 6f 6e 74 65 6e 74 2d 4c 65 6e 67 74 68 3a 20 Content-Length: 
0010: 30 0d 0a                                        0..
<= Recv header, 19 bytes (0x13)
0000: 43 6f 6e 6e 65 63 74 69 6f 6e 3a 20 63 6c 6f 73 Connection: clos
0010: 65 0d 0a                                        e..
<= Recv header, 2 bytes (0x2)
0000: 0d 0a                                           ..
== Info: Closing connection 0
```

But nothing happens. Expection is that the power of my plug should be switch on.
The content of the script is also not updated.

From the code

```
      Webserver->on("/exs", HTTP_POST,[]() { Webserver->sendHeader("Location","/exs");Webserver->send(303);}, script_upload_start);
```

it seems to me that the response is okay, however what is the 'script_upload_start' function expecting ?

**Comment by gemu2015** [comment]
as stated above my external editor uploads scripts that way. here is a wireshark tracing how this works in detail.
however i could not get curl to mimic this behavior.
[upload_esp32.pcapng.zip](https://github.com/arendst/Tasmota/files/7868393/upload_esp32.pcapng.zip)

**Comment by barbudor** [comment, marked as answer]
Thanks @gemu2015 
I tried on the `/ta` URL which is the one used by the embedded editor.
But despite even making a python script to exactly have the same data between the embedded editor and my script, something blocks with my scripts.
I will try the `/exs` URL following your trace.

BY the way, I have not seen any reference to your editor in the Scripting-Language page. Is that something you could share ?
Thanks

**Reply by gemu2015** [reply]
from the docs:

Optional external editor~

you may use a special external editor with syntax highlighting to edit the scripts. (mac and pc) you may use any number of comments and indents to make it better readable. then with cmd r the script is transfered to the ESP and immediately started. (all comments and indents are removed before transfering) see further info and download here

https://www.dropbox.com/sh/0us18ohui4c3k82/AACcVmpZ4AfpdrWE_MPFGmbma?dl=0
